topdown/graphql.go

```diff
@@ -340,29 +435,48 @@ func builtinGraphQLParseSchema(_ BuiltinContext, operands []*ast.Term, iter func
 }
 
 func builtinGraphQLIsValid(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
-	// Get the raw strings from each operand so that we can
-	// feed them to the GraphQL parser functions.
-	rawQuery, err := builtins.StringOperand(operands[0].Value, 1)
-	if err != nil {
-		return iter(ast.BooleanTerm(false))
-	}
-	rawSchema, err := builtins.StringOperand(operands[1].Value, 1)
-	if err != nil {
+	var queryDoc *gqlast.QueryDocument
+	var schemaDoc *gqlast.SchemaDocument
+	var err error
+
+	switch x := operands[0].Value.(type) {
+	case ast.String:
+		queryDoc, err = parseQuery(string(x))
+		if err != nil {
+			return iter(ast.BooleanTerm(false))
+		}
+	case ast.Object:
+		queryDoc, err = objectToQueryDocument(x)
+		if err != nil {
+			return iter(ast.BooleanTerm(false))
+		}
+	default:
+		// Error if wrong type.
 		return iter(ast.BooleanTerm(false))
 	}
```

Comment:
Perhaps helpers of some sort would help?
```go
func valueToQuery(v ast.Value) *gqlast.QueryDocument { /* ... */ }
func valueToSchema(v ast.Value) *gqlast.SchemaDocument { /* ... */ }
```
so we could do
```go
query := valueToQuery(operands[0].Value)
if query == nil {
    return iter(ast.BooleanTerm(false))
}
schema := valueToSchema(operands[1].Value)
if schema == nil {
    return iter(ast.BooleanTerm(false))
}

// the rest
````

Reply:
Returning an error too might be worthwhile if we bury the validation logic into those helpers in the future. :thinking: